File: research/src/game2048/game.py
"""ゲーム進行 (SPEC.md #10)。web/src/game/game.ts に対応。"""

from __future__ import annotations

from dataclasses import dataclass

from .board import Board, create_empty_board, get_empty_cells
from .move import Direction, is_game_over, move
from .rng import Rng

TILE_2_PROBABILITY = 0.9


@dataclass
class GameState:
    board: Board
    score: int
    move_count: int
    game_over: bool


def spawn_random_tile(board: Board, rng: Rng) -> Board:
    """空きマスに新しいタイルを1つ生成する。2:90%, 4:10% (SPEC.md #10.4)"""
    empty_cells = get_empty_cells(board)
    if not empty_cells:
        return board

    cell_index = empty_cells[int(rng.next() * len(empty_cells))]
    value = 2 if rng.next() < TILE_2_PROBABILITY else 4

    next_board = board.copy()
    next_board[cell_index] = value
    return next_board


def create_initial_state(rng: Rng) -> GameState:
    """初期盤面（タイル2つ配置済み）を持つ GameState を生成する"""
    board = create_empty_board()
    board = spawn_random_tile(board, rng)
    board = spawn_random_tile(board, rng)
    return GameState(board=board, score=0, move_count=0, game_over=False)


def apply_move(state: GameState, direction: Direction, rng: Rng) -> GameState:
    """手を1つ適用する。盤面が変化しない方向を指定した場合は state をそのまま返す。"""
    if state.game_over:
        return state

    result = move(state.board, direction)
    if not result.moved:
        return state

    board_with_new_tile = spawn_random_tile(result.board, rng)
    return GameState(
        board=board_with_new_tile,
        score=state.score + result.score_delta,
        move_count=state.move_count + 1,
        game_over=is_game_over(board_with_new_tile),
    )

File: web/src/game/game.ts
import { createEmptyBoard, DEFAULT_BOARD_SIZE, getEmptyCells } from "./board";
import { isGameOver, move } from "./move";
import type { Rng } from "./rng";
import type { Board, Direction, GameState } from "./types";

/** 空きマスに新しいタイルを1つ生成する。2:90%, 4:10% (SPEC.md #10.4) */
export function spawnRandomTile(board: Board, rng: Rng): Board {
  const emptyCells = getEmptyCells(board);
  if (emptyCells.length === 0) return board;

  const cellIndex = emptyCells[Math.floor(rng.next() * emptyCells.length)];
  const value = rng.next() < 0.9 ? 2 : 4;

  const next = board.slice();
  next[cellIndex] = value;
  return next;
}

/** 初期盤面（タイル2つ配置済み）を持つ GameState を生成する。盤面サイズは可変 (issue #16, #17) */
export function createInitialState(rng: Rng, boardSize: number = DEFAULT_BOARD_SIZE): GameState {
  let board = createEmptyBoard(boardSize);
  board = spawnRandomTile(board, rng);
  board = spawnRandomTile(board, rng);

  return {
    board,
    score: 0,
    moveCount: 0,
    gameOver: false,
  };
}

/**
 * 手を1つ適用する。盤面が変化しない方向を指定した場合は state をそのまま返す。
 */
export function applyMove(state: GameState, direction: Direction, rng: Rng): GameState {
  if (state.gameOver) return state;

  const result = move(state.board, direction);
  if (!result.moved) return state;

  const boardWithNewTile = spawnRandomTile(result.board, rng);

  return {
    board: boardWithNewTile,
    score: state.score + result.scoreDelta,
    moveCount: state.moveCount + 1,
    gameOver: isGameOver(boardWithNewTile),
  };
}

export { getEmptyCells, getMaxTile, cloneBoard } from "./board";
export { getValidMoves, isGameOver, move } from "./move";
